executor/prepared.go

```diff
@@ -0,0 +1,279 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/executor/converter"
+	"github.com/pingcap/tidb/field"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/optimizer"
+	"github.com/pingcap/tidb/optimizer/evaluator"
+	"github.com/pingcap/tidb/optimizer/plan"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/stmt"
+)
+
+type paramMarkerSorter struct {
+	markers []*ast.ParamMarkerExpr
+}
+
+func (p *paramMarkerSorter) Len() int {
+	return len(p.markers)
+}
+
+func (p *paramMarkerSorter) Less(i, j int) bool {
+	return p.markers[i].Offset < p.markers[j].Offset
+}
+
+func (p *paramMarkerSorter) Swap(i, j int) {
+	p.markers[i], p.markers[j] = p.markers[j], p.markers[i]
+}
+
+type paramMarkerExtractor struct {
+	markers []*ast.ParamMarkerExpr
+}
+
+func (e *paramMarkerExtractor) Enter(in ast.Node) (ast.Node, bool) {
+	return in, false
+}
+
+func (e *paramMarkerExtractor) Leave(in ast.Node) (ast.Node, bool) {
+	if x, ok := in.(*ast.ParamMarkerExpr); ok {
+		e.markers = append(e.markers, x)
+	}
+	return in, true
+}
+
+// Prepared represents a prepared statement.
+type Prepared struct {
+	Stmt          ast.StmtNode
+	Params        []*ast.ParamMarkerExpr
+	SchemaVersion int64
+}
+
+// PrepareExec represents a PREPARE executor.
+type PrepareExec struct {
+	IS      infoschema.InfoSchema
+	Ctx     context.Context
+	Name    string
+	SQLText string
+
+	ID           uint32
+	ResultFields []*field.ResultField
+	ParamCount   int
+	Err          error
+}
+
+// Fields implements Executor Fields interface.
+func (e *PrepareExec) Fields() []*ast.ResultField {
+	// returns nil to indicate prepare will not return Recordset.
+	return nil
+}
+
+// Next implements Executor Next interface.
+func (e *PrepareExec) Next() (*Row, error) {
+	e.DoPrepare()
+	return nil, e.Err
+}
+
+// Close implements plan.Plan Close interface.
+func (e *PrepareExec) Close() error {
+	return nil
+}
+
+// DoPrepare prepares the statement, it can be called multiple times without
+// side effect.
+func (e *PrepareExec) DoPrepare() {
+	vars := variable.GetSessionVars(e.Ctx)
+	if e.ID != 0 {
+		_, ok := vars.PreparedStmts[e.ID]
```

Comment:
If the ID is not 0, that must be the case when we retry a prepared statement.
The other fields should be already set.

Here we make sure retry a prepared statement is idempotent.

There is no way to do a new prepare with an ID.